std/sumtype.d

```diff
@@ -0,0 +1,2488 @@
+        ~this()
+        {
+            this.match!destroyIfOwner;
+        }
+    }
+
+    invariant
+    {
+        this.match!((ref value) {
+            static if (is(typeof(value) == class))
+            {
+                if (value !is null)
+                {
+                    assert(value);
+                }
+            }
+            else static if (is(typeof(value) == struct))
+            {
+                assert(&value);
+            }
+        });
+    }
+
+    // Workaround for dlang issue 21400
+    version (StdDdoc)
+    {
+        /**
+         * Returns a string representation of the `SumType`'s current value.
+         *
+         * Not available when compiled with `-betterC`.
+         */
+        string toString(this This)();
+
+        /**
+         * Handles formatted writing of the `SumType`'s current value.
+         *
+         * Not available when compiled with `-betterC`.
+         *
+         * Params:
+         *   sink = Output range to write to.
+         *   fmt = Format specifier to use.
+         *
+         * See_Also: $(REF formatValue, std,format)
+         */
+        void toString(this This, Sink, Char)(ref Sink sink, const ref FormatSpec!Char fmt);
+    }
+
+    version (D_BetterC) {} else
+    /**
+     * Returns a string representation of the `SumType`'s current value.
+     *
+     * Not available when compiled with `-betterC`.
+     */
+    string toString(this This)()
+    {
+        import std.conv : to;
+
+        return this.match!(to!string);
+    }
+
+    version (D_BetterC) {} else
+    /**
+     * Handles formatted writing of the `SumType`'s current value.
+     *
+     * Not available when compiled with `-betterC`.
+     *
+     * Params:
+     *   sink = Output range to write to.
+     *   fmt = Format specifier to use.
+     *
+     * See_Also: $(REF formatValue, std,format)
+     */
+    void toString(this This, Sink, Char)(ref Sink sink, const ref FormatSpec!Char fmt)
+    {
+        import std.format : formatValue;
+
+        this.match!((ref value) {
+            formatValue(sink, value, fmt);
+        });
+    }
+
+    static if (allSatisfy!(isHashable, Map!(ConstOf, Types)))
+    {
+        // Workaround for dlang issue 21400
+        version (StdDdoc)
+        {
+            /**
+             * Returns the hash of the `SumType`'s current value.
+             *
+             * Not available when compiled with `-betterC`.
+             */
+            size_t toHash() const;
+        }
+
+        // Workaround for dlang issue 20095
+        version (D_BetterC) {} else
+        /**
+         * Returns the hash of the `SumType`'s current value.
+         *
+         * Not available when compiled with `-betterC`.
+         */
+        size_t toHash() const
+        {
+            return this.match!hashOf;
+        }
+    }
+}
+
+// Construction
+@safe unittest
+{
+    alias MySum = SumType!(int, float);
+
+    assert(__traits(compiles, MySum(42)));
```

Comment:
Wouldn't it be better to just write the code instead of asserting it compiles? If it ever fails to, the compiler will spit out an error message.

Reply:
The `assert` communicates intent better (IMO) and avoids masking other compile-time errors in case of a failed `static assert` (e.g., in `match`).

Reply:
I don't think it makes much of a difference to the maintainer - if this fails, it needs to be fixed pronto and that's the end of it. FWIW too many errors can be move overwhelming than informative. I forgot how long building vs. running the unittests takes, if the static assert causes the process to fail a lot sooner it has an advantage.

Reply:
`sumtype`'s test suite builds quickly enough (~2s on my machine) that I don't think we need to worry too much about this kind of optimization. I suppose if someone is building the entire Phobos test suite sequentially it could help, but I don't expect that's a very common operation, and the error-masking issue is a real annoyance during normal maintenance work (I speak from experience).